taskchampion/src/server/cloud/aws.rs

```diff
@@ -0,0 +1,628 @@
+
+        // Clean up.
+        svc.del(&pfx("testy")).unwrap();
+    }
+
+    #[test]
+    fn compare_and_swap_expected_no_file() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+
+        svc.put(&pfx("testy"), b"foo1").unwrap();
+        assert!(!svc
+            .compare_and_swap(&pfx("testy"), None, b"bar".to_vec())
+            .unwrap());
+        let got = svc.get(&pfx("testy")).unwrap();
+        assert_eq!(got, Some(b"foo1".to_vec()));
+
+        // Clean up.
+        svc.del(&pfx("testy")).unwrap();
+    }
+
+    #[test]
+    fn compare_and_swap_mismatch() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+
+        // Create the existing file, with two generations.
+        svc.put(&pfx("testy"), b"foo1").unwrap();
+        svc.put(&pfx("testy"), b"foo2").unwrap();
+        assert!(!svc
+            .compare_and_swap(&pfx("testy"), Some(b"foo1".to_vec()), b"bar".to_vec())
+            .unwrap());
+        let got = svc.get(&pfx("testy")).unwrap();
+        assert_eq!(got, Some(b"foo2".to_vec()));
+
+        // Clean up.
+        svc.del(&pfx("testy")).unwrap();
+    }
+
+    #[test]
+    fn compare_and_swap_changes() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+
+        // Create the existing file, but since it is named "racing-put" it will change just
+        // before the `put_object` call.
+        svc.put(&pfx("racing-put"), b"foo1").unwrap();
+        assert!(!svc
+            .compare_and_swap(&pfx("racing-put"), Some(b"foo1".to_vec()), b"bar".to_vec())
+            .unwrap());
+        let got = svc.get(&pfx("racing-put")).unwrap();
+        assert_eq!(got, Some(b"CHANGED".to_vec()));
+    }
+
+    #[test]
+    fn compare_and_swap_disappears() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+
+        // Create the existing file, but since it is named "racing-delete" it will disappear just
+        // before the `put_object` call.
+        svc.put(&pfx("racing-delete"), b"foo1").unwrap();
+        assert!(!svc
+            .compare_and_swap(
+                &pfx("racing-delete"),
+                Some(b"foo1".to_vec()),
+                b"bar".to_vec()
+            )
+            .unwrap());
+        let got = svc.get(&pfx("racing-delete")).unwrap();
+        assert_eq!(got, None);
+    }
+
+    #[test]
+    fn compare_and_swap_appears() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+
+        // Create the existing file, but since it is named "racing-put" the file will appear just
+        // before the `put_object` call.
+        assert!(!svc
+            .compare_and_swap(&pfx("racing-put"), None, b"bar".to_vec())
+            .unwrap());
+        let got = svc.get(&pfx("racing-put")).unwrap();
+        assert_eq!(got, Some(b"CHANGED".to_vec()));
+    }
```

Comment:
I don't understand the name of this test and it isn't clear to me how the behavior tested is distinct from that of `compare_and_swap_changes`.

Reply:
I've added some comments.